From 587113aaad652e8e93a6305b01c216e4dfab410e Mon Sep 17 00:00:00 2001
From: Kevin Peizner <kevin.peizner@racepointenergy.com>
Date: Wed, 28 Oct 2020 11:51:55 -0700
Subject: [PATCH] Make dispatchConfig.cmake not depend on build dir

The default dispatchConfig.cmake generated file depends on the
generated dispatchExports.cmake file via a build dir specific path.

Furthermore, the original dispatchExports.cmake contains build dir
specific paths.

This breaks inclusion in other projects unless said project has access
to the build & src tree of libdispatch.

This commit changes the dispatchConfig.cmake & dispatchExports.cmake
files so that the paths within them are based relative to the
installation of libdispatch.
---
 cmake/modules/CMakeLists.txt          | 24 +++++++++++++++++++-----
 cmake/modules/dispatchConfig.cmake.in |  8 +++++++-
 src/BlocksRuntime/CMakeLists.txt      |  3 ++-
 src/CMakeLists.txt                    |  9 +++++----
 src/swift/CMakeLists.txt              |  6 ++++--
 5 files changed, 37 insertions(+), 13 deletions(-)

diff --git a/cmake/modules/CMakeLists.txt b/cmake/modules/CMakeLists.txt
index 10cc0e1..33df058 100644
--- a/cmake/modules/CMakeLists.txt
+++ b/cmake/modules/CMakeLists.txt
@@ -1,7 +1,21 @@
 
-set(DISPATCH_EXPORTS_FILE ${CMAKE_CURRENT_BINARY_DIR}/dispatchExports.cmake)
-configure_file(dispatchConfig.cmake.in
-  ${CMAKE_CURRENT_BINARY_DIR}/dispatchConfig.cmake)
+set(INCLUDE_INSTALL_DIR lib/swift)
+set(LIB_INSTALL_DIR lib/swift)
+
+include(CMakePackageConfigHelpers)
+configure_package_config_file(dispatchConfig.cmake.in
+  ${CMAKE_CURRENT_BINARY_DIR}/dispatchConfig.cmake
+  INSTALL_DESTINATION ${LIB_INSTALL_DIR}/dispatch/cmake
+  PATH_VARS INCLUDE_INSTALL_DIR LIB_INSTALL_DIR)
+
+#write_basic_package_version_file(
+#  ${CMAKE_CURRENT_BINARY_DIR}/dispatchConfigVersion.cmake
+#  VERSION 1.2.3
+#  COMPATIBILITY SameMajorVersion )
+
+install(EXPORT dispatchExports DESTINATION ${LIB_INSTALL_DIR}/dispatch/cmake)
+install(FILES
+        ${CMAKE_CURRENT_BINARY_DIR}/dispatchConfig.cmake
+        DESTINATION ${LIB_INSTALL_DIR}/dispatch/cmake )
+
 
-get_property(DISPATCH_EXPORTS GLOBAL PROPERTY DISPATCH_EXPORTS)
-export(TARGETS ${DISPATCH_EXPORTS} FILE ${DISPATCH_EXPORTS_FILE})
diff --git a/cmake/modules/dispatchConfig.cmake.in b/cmake/modules/dispatchConfig.cmake.in
index 81228f2..745710e 100644
--- a/cmake/modules/dispatchConfig.cmake.in
+++ b/cmake/modules/dispatchConfig.cmake.in
@@ -1,7 +1,13 @@
 
 set(DISPATCH_HAS_SWIFT_SDK_OVERLAY @ENABLE_SWIFT@)
 
+@PACKAGE_INIT@
+set_and_check(DISPATCH_INCLUDE_DIR "@PACKAGE_INCLUDE_INSTALL_DIR@")
+set_and_check(DISPATCH_LIB_DIR "@PACKAGE_LIB_INSTALL_DIR@")
+
+check_required_components(dispatch)
+
 if(NOT TARGET dispatch)
-  include(@DISPATCH_EXPORTS_FILE@)
+  include(${DISPATCH_LIB_DIR}/dispatch/cmake/dispatchExports.cmake)
 endif()
 
diff --git a/src/BlocksRuntime/CMakeLists.txt b/src/BlocksRuntime/CMakeLists.txt
index 1bed202..9f5822b 100644
--- a/src/BlocksRuntime/CMakeLists.txt
+++ b/src/BlocksRuntime/CMakeLists.txt
@@ -13,7 +13,8 @@ if(CMAKE_SYSTEM_NAME STREQUAL Windows)
 endif()
 
 target_include_directories(BlocksRuntime PUBLIC
-  ${CMAKE_CURRENT_SOURCE_DIR})
+    $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}>
+    $<INSTALL_INTERFACE:lib/swift/Block>)
 if(HAVE_OBJC AND CMAKE_DL_LIBS)
   target_link_libraries(BlocksRuntime PUBLIC
     ${CMAKE_DL_LIBS})
diff --git a/src/CMakeLists.txt b/src/CMakeLists.txt
index adc989d..8d38c14 100644
--- a/src/CMakeLists.txt
+++ b/src/CMakeLists.txt
@@ -96,10 +96,11 @@ set_target_properties(dispatch PROPERTIES
   POSITION_INDEPENDENT_CODE YES)
 
 target_include_directories(dispatch PUBLIC
-  ${PROJECT_BINARY_DIR}
-  ${PROJECT_SOURCE_DIR}
-  ${CMAKE_CURRENT_SOURCE_DIR}
-  ${CMAKE_CURRENT_BINARY_DIR})
+    $<BUILD_INTERFACE:${PROJECT_BINARY_DIR}>
+    $<BUILD_INTERFACE:${PROJECT_SOURCE_DIR}>
+    $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}>
+    $<BUILD_INTERFACE:${CMAKE_CURRENT_BINARY_DIR}>
+    $<INSTALL_INTERFACE:lib/swift>)
 target_include_directories(dispatch PRIVATE
   ${PROJECT_SOURCE_DIR}/private)
 
diff --git a/src/swift/CMakeLists.txt b/src/swift/CMakeLists.txt
index 5392472..2c7de93 100644
--- a/src/swift/CMakeLists.txt
+++ b/src/swift/CMakeLists.txt
@@ -34,8 +34,10 @@ target_compile_options(swiftDispatch PRIVATE
   "SHELL:-Xcc -I${PROJECT_SOURCE_DIR}")
 set_target_properties(swiftDispatch PROPERTIES
   Swift_MODULE_NAME Dispatch
-  Swift_MODULE_DIRECTORY ${CMAKE_CURRENT_BINARY_DIR}/swift
-  INTERFACE_INCLUDE_DIRECTORIES ${CMAKE_CURRENT_BINARY_DIR}/swift)
+  Swift_MODULE_DIRECTORY ${CMAKE_CURRENT_BINARY_DIR}/swift)
+target_include_directories(swiftDispatch INTERFACE
+    $<BUILD_INTERFACE:${CMAKE_CURRENT_BINARY_DIR}/swift>
+    $<INSTALL_INTERFACE:lib/swift>)
 target_link_libraries(swiftDispatch PRIVATE
   DispatchStubs
   BlocksRuntime::BlocksRuntime)
